File: apps/wallet-mobile/etrid-wallet/services/DAOService.ts
/**
 * DAO Service
 * Handles DAO operations, creation, and membership management
 */

import { DAO, DAODetail, DAOParams, DAOMember, DAOStats } from '@/types/dao';

export class DAOService {
  private daos: Map<string, DAO> = new Map();
  private userMemberships: Map<string, string[]> = new Map(); // userId -> daoIds

  /**
   * Create a new DAO
   */
  async createDAO(params: DAOParams): Promise<DAO> {
    // Validate params
    if (!params.name || !params.description) {
      throw new Error('Name and description are required');
    }

    // Generate treasury address (would be actual smart contract deployment)
    const treasuryAddress = this.generateTreasuryAddress();

    const dao: DAO = {
      id: Date.now().toString(),
      name: params.name,
      description: params.description,
      logoUrl: params.logoUrl,
      governance: params.governance,
      treasuryAddress,
      memberCount: params.initialMembers?.length || 1,
      activeProposalCount: 0,
      treasuryValue: params.initialTreasuryAmount || '0',
      createdAt: new Date(),
      userRole: 'owner',
    };

    this.daos.set(dao.id, dao);

    // Add creator as owner
    const currentUserId = await this.getCurrentUserId();
    this.addUserMembership(currentUserId, dao.id);

    // Add initial members
    if (params.initialMembers) {
      for (const memberId of params.initialMembers) {
        this.addUserMembership(memberId, dao.id);
      }
    }

    // Persist to storage
    await this.saveDAOs();

    // Deploy DAO smart contract (TODO: implement actual deployment)
    // await this.deployDAOContract(dao);

    return dao;
  }

  /**
   * Get user's DAOs with optional filter
   */
  async getDAOs(filter?: 'member' | 'owner'): Promise<DAO[]> {
    const userId = await this.getCurrentUserId();
    const userDaoIds = this.userMemberships.get(userId) || [];

    let daos = Array.from(this.daos.values()).filter((dao) =>
      userDaoIds.includes(dao.id)
    );

    if (filter === 'owner') {
      daos = daos.filter((dao) => dao.userRole === 'owner');
    } else if (filter === 'member') {
      daos = daos.filter((dao) => dao.userRole === 'member' || dao.userRole === 'voter');
    }

    return daos;
  }

  /**
   * Get detailed DAO information
   */
  async getDAODetails(daoId: string): Promise<DAODetail> {
    const dao = this.daos.get(daoId);
    if (!dao) {
      throw new Error('DAO not found');
    }

    // Fetch additional details
    const members = await this.getDAOMembers(daoId);
    const proposals = await this.getDAOProposals(daoId);
    const treasury = await this.getDAOTreasury(daoId);
    const stats = await this.getDAOStats(daoId);

    const daoDetail: DAODetail = {
      ...dao,
      members,
      proposals,
      treasury,
      stats,
    };

    return daoDetail;
  }

  /**
   * Join a DAO
   */
  async joinDAO(daoId: string): Promise<void> {
    const dao = this.daos.get(daoId);
    if (!dao) {
      throw new Error('DAO not found');
    }

    const userId = await this.getCurrentUserId();

    // Check if already a member
    const userDaoIds = this.userMemberships.get(userId) || [];
    if (userDaoIds.includes(daoId)) {
      throw new Error('Already a member of this DAO');
    }

    // Check membership requirements
    const canJoin = await this.checkMembershipRequirements(dao, userId);
    if (!canJoin) {
      throw new Error('Does not meet membership requirements');
    }

    // Add membership
    this.addUserMembership(userId, daoId);

    // Update member count
    dao.memberCount += 1;
    this.daos.set(daoId, dao);

    await this.saveDAOs();

    // TODO: Call smart contract to register membership
    // await this.registerMemberOnChain(daoId, userId);
  }

  /**
   * Leave a DAO
   */
  async leaveDAO(daoId: string): Promise<void> {
    const dao = this.daos.get(daoId);
    if (!dao) {
      throw new Error('DAO not found');
    }

    // Cannot leave if you're the owner
    if (dao.userRole === 'owner') {
      throw new Error('Owner cannot leave the DAO');
    }

    const userId = await this.getCurrentUserId();
    this.removeUserMembership(userId, daoId);

    // Update member count
    dao.memberCount = Math.max(0, dao.memberCount - 1);
    this.daos.set(daoId, dao);

    await this.saveDAOs();

    // TODO: Call smart contract to remove membership
    // await this.removeMemberOnChain(daoId, userId);
  }

  /**
   * Get DAO members
   */
  private async getDAOMembers(daoId: string): Promise<DAOMember[]> {
    // TODO: Fetch from API or blockchain
    return this.getMockMembers(daoId);
  }

  /**
   * Get DAO proposals
   */
  private async getDAOProposals(daoId: string): Promise<any[]> {
    // This would be implemented in DAOProposalService
    return [];
  }

  /**
   * Get DAO treasury
   */
  private async getDAOTreasury(daoId: string): Promise<any> {
    // This would be implemented in DAOTreasuryService
    return {
      daoId,
      totalValue: '0',
      assets: [],
      transactions: [],
      analytics: {},
    };
  }

  /**
   * Get DAO statistics
   */
  private async getDAOStats(daoId: string): Promise<DAOStats> {
    // TODO: Calculate from actual data
    return {
      totalProposals: 0,
      passedProposals: 0,
      rejectedProposals: 0,
      averageParticipation: 0,
      treasuryGrowth: 0,
      activeMembers: 0,
    };
  }

  /**
   * Check if user meets membership requirements
   */
  private async checkMembershipRequirements(
    dao: DAO,
    userId: string
  ): Promise<boolean> {
    const { membershipType, tokenAddress, nftAddress } = dao.governance;

    switch (membershipType) {
      case 'open':
        return true;

      case 'token-gated':
        if (!tokenAddress) return false;
        // Check if user holds required tokens
        const tokenBalance = await this.getUserTokenBalance(userId, tokenAddress);
        return parseFloat(tokenBalance) > 0;

      case 'nft-gated':
        if (!nftAddress) return false;
        // Check if user owns the required NFT
        const ownsNFT = await this.checkUserOwnsNFT(userId, nftAddress);
        return ownsNFT;

      case 'invite-only':
        // Would check if user has been invited
        return false;

      default:
        return false;
    }
  }

  /**
   * Add user membership
   */
  private addUserMembership(userId: string, daoId: string): void {
    const userDaoIds = this.userMemberships.get(userId) || [];
    if (!userDaoIds.includes(daoId)) {
      userDaoIds.push(daoId);
      this.userMemberships.set(userId, userDaoIds);
    }
  }

  /**
   * Remove user membership
   */
  private removeUserMembership(userId: string, daoId: string): void {
    const userDaoIds = this.userMemberships.get(userId) || [];
    const filtered = userDaoIds.filter((id) => id !== daoId);
    this.userMemberships.set(userId, filtered);
  }

  /**
   * Generate treasury address
   */
  private generateTreasuryAddress(): string {
    // TODO: Deploy actual treasury contract
    return '0x' + Math.random().toString(16).substring(2, 42);
  }

  // Storage methods
  private async saveDAOs(): Promise<void> {
    const daosArray = Array.from(this.daos.values());
    // TODO: Persist to AsyncStorage or API
    // await AsyncStorage.setItem('daos', JSON.stringify(daosArray));
  }

  private async loadDAOs(): Promise<void> {
    // TODO: Load from AsyncStorage or API
    // const daosJson = await AsyncStorage.getItem('daos');
    // if (daosJson) {
    //   const daos = JSON.parse(daosJson);
    //   daos.forEach((dao: DAO) => {
    //     this.daos.set(dao.id, dao);
    //   });
    // }
  }

  // Mock/Helper methods
  private async getCurrentUserId(): Promise<string> {
    // TODO: Get from auth service
    return 'user-1';
  }

  private async getUserTokenBalance(userId: string, tokenAddress: string): Promise<string> {
    // TODO: Check actual token balance
    return '100';
  }

  private async checkUserOwnsNFT(userId: string, nftAddress: string): Promise<boolean> {
    // TODO: Check actual NFT ownership
    return false;
  }

  private getMockMembers(daoId: string): DAOMember[] {
    return [
      {
        id: '1',
        userId: 'user-1',
        address: '0x1234...',
        username: 'Alice',
        role: 'owner',
        votingPower: 1000,
        joinedAt: new Date(),
        proposalsCreated: 5,
        votesCast: 12,
      },
      {
        id: '2',
        userId: 'user-2',
        address: '0x5678...',
        username: 'Bob',
        role: 'member',
        votingPower: 500,
        joinedAt: new Date(),
        proposalsCreated: 2,
        votesCast: 8,
      },
    ];
  }
}

export const daoService = new DAOService();

File: apps/EtridWalletApp/EtridWallet/Services/DAOService.swift
//
//  DAOService.swift
//  EtridWallet
//
//  Service for DAO operations and governance
//

import Foundation
import Combine

@MainActor
class DAOService: ObservableObject {
    static let shared = DAOService()

    @Published var daos: [DAO] = []
    @Published var proposals: [String: [DAOProposal]] = [:] // daoId -> proposals
    @Published var isLoading = false
    @Published var error: Error?

    private let storageKey = "daos"
    private let proposalsKey = "dao_proposals"

    private init() {
        loadDAOs()
        loadProposals()
    }

    // MARK: - DAO Management

    func createDAO(_ params: CreateDAOParams) async throws -> DAO {
        isLoading = true
        defer { isLoading = false }

        // Validate parameters
        guard !params.name.isEmpty, !params.description.isEmpty else {
            throw WalletError.invalidInput("Name and description are required")
        }

        guard params.quorumPercentage > 0 && params.quorumPercentage <= 100 else {
            throw WalletError.invalidInput("Quorum must be between 0 and 100")
        }

        // Generate treasury address (would deploy actual contract)
        let treasuryAddress = generateTreasuryAddress()

        let dao = DAO(
            id: UUID().uuidString,
            name: params.name,
            description: params.description,
            logoUrl: params.logoUrl,
            governance: params.governance,
            treasuryAddress: treasuryAddress,
            memberCount: 1 + (params.initialMembers?.count ?? 0),
            activeProposalCount: 0,
            treasuryValue: params.initialTreasuryAmount ?? "0",
            createdAt: Date(),
            userRole: .owner
        )

        daos.append(dao)
        saveDAOs()

        // TODO: Deploy DAO smart contract
        // let contractAddress = try await deployDAOContract(dao)

        return dao
    }

    func getDAOs(filter: DAO.MemberRole? = nil) -> [DAO] {
        if let filter = filter {
            return daos.filter { $0.userRole == filter }
        }
        return daos
    }

    func getDAO(id: String) -> DAO? {
        daos.first { $0.id == id }
    }

    func joinDAO(id: String) async throws {
        guard var dao = getDAO(id: id) else {
            throw WalletError.unknown("DAO not found")
        }

        // Check membership requirements
        let canJoin = try await checkMembershipRequirements(dao)
        guard canJoin else {
            throw WalletError.unknown("Does not meet membership requirements")
        }

        // Update DAO
        dao.memberCount += 1
        dao.userRole = .member
        updateDAO(dao)

        // TODO: Register membership on-chain
    }

    func leaveDAO(id: String) async throws {
        guard var dao = getDAO(id: id) else {
            throw WalletError.unknown("DAO not found")
        }

        guard dao.userRole != .owner else {
            throw WalletError.unknown("Owner cannot leave DAO")
        }

        dao.memberCount = max(0, dao.memberCount - 1)
        updateDAO(dao)

        // TODO: Remove membership on-chain
    }

    // MARK: - Proposal Management

    func createProposal(_ params: CreateProposalParams) async throws -> DAOProposal {
        guard let dao = getDAO(id: params.daoId) else {
            throw WalletError.unknown("DAO not found")
        }

        isLoading = true
        defer { isLoading = false }

        let proposal = DAOProposal(
            id: UUID().uuidString,
            daoId: params.daoId,
            title: params.title,
            description: params.description,
            proposer: try await getCurrentAddress(),
            proposerName: "You",
            actions: params.actions,
            createdAt: Date(),
            startTime: Date(),
            endTime: Date().addingTimeInterval(TimeInterval(params.votingPeriodDays * 86400)),
            votesFor: "0",
            votesAgainst: "0",
            votesAbstain: "0",
            status: .active,
            userVote: nil
        )

        var daoProposals = proposals[params.daoId] ?? []
        daoProposals.append(proposal)
        proposals[params.daoId] = daoProposals

        saveProposals()

        // Update DAO
        var updatedDAO = dao
        updatedDAO.activeProposalCount += 1
        updateDAO(updatedDAO)

        // TODO: Submit proposal to smart contract

        return proposal
    }

    func getProposals(daoId: String, filter: DAOProposal.ProposalStatus? = nil) -> [DAOProposal] {
        let daoProposals = proposals[daoId] ?? []
        if let filter = filter {
            return daoProposals.filter { $0.status == filter }
        }
        return daoProposals
    }

    func getProposal(id: String, daoId: String) -> DAOProposal? {
        proposals[daoId]?.first { $0.id == id }
    }

    func vote(proposalId: String, daoId: String, voteType: DAOVoteType, votingPower: String) async throws {
        guard var proposal = getProposal(id: proposalId, daoId: daoId) else {
            throw WalletError.unknown("Proposal not found")
        }

        guard proposal.status == .active else {
            throw WalletError.invalidInput("Proposal is not active")
        }

        guard proposal.userVote == nil else {
            throw WalletError.invalidInput("Already voted on this proposal")
        }

        isLoading = true
        defer { isLoading = false }

        // Update votes
        switch voteType {
        case .for:
            let current = Double(proposal.votesFor) ?? 0
            let power = Double(votingPower) ?? 0
            proposal.votesFor = String(current + power)
        case .against:
            let current = Double(proposal.votesAgainst) ?? 0
            let power = Double(votingPower) ?? 0
            proposal.votesAgainst = String(current + power)
        case .abstain:
            let current = Double(proposal.votesAbstain) ?? 0
            let power = Double(votingPower) ?? 0
            proposal.votesAbstain = String(current + power)
        }

        proposal.userVote = voteType
        updateProposal(proposal, daoId: daoId)

        // TODO: Submit vote to smart contract
    }

    func executeProposal(proposalId: String, daoId: String) async throws {
        guard var proposal = getProposal(id: proposalId, daoId: daoId) else {
            throw WalletError.unknown("Proposal not found")
        }

        guard proposal.status == .succeeded else {
            throw WalletError.invalidInput("Proposal has not succeeded")
        }

        isLoading = true
        defer { isLoading = false }

        proposal.status = .executed
        updateProposal(proposal, daoId: daoId)

        // TODO: Execute proposal actions on-chain
    }

    // MARK: - Treasury

    func getTreasury(daoId: String) async throws -> DAOTreasury {
        // TODO: Fetch treasury data from blockchain
        return DAOTreasury(
            daoId: daoId,
            totalValue: "0",
            assets: [],
            transactions: [],
            analytics: nil
        )
    }

    // MARK: - Members

    func getMembers(daoId: String) async throws -> [DAOMember] {
        // TODO: Fetch members from smart contract
        return getMockMembers()
    }

    func addMember(daoId: String, address: String) async throws {
        guard var dao = getDAO(id: daoId) else {
            throw WalletError.unknown("DAO not found")
        }

        guard dao.userRole == .owner || dao.userRole == .admin else {
            throw WalletError.unknown("Insufficient permissions")
        }

        dao.memberCount += 1
        updateDAO(dao)

        // TODO: Add member on-chain
    }

    func removeMember(daoId: String, address: String) async throws {
        guard var dao = getDAO(id: daoId) else {
            throw WalletError.unknown("DAO not found")
        }

        guard dao.userRole == .owner || dao.userRole == .admin else {
            throw WalletError.unknown("Insufficient permissions")
        }

        dao.memberCount = max(0, dao.memberCount - 1)
        updateDAO(dao)

        // TODO: Remove member on-chain
    }

    // MARK: - Stats

    func getStats(daoId: String) async throws -> DAOStats {
        let daoProposals = proposals[daoId] ?? []
        let total = daoProposals.count
        let passed = daoProposals.filter { $0.status == .succeeded || $0.status == .executed }.count
        let rejected = daoProposals.filter { $0.status == .defeated }.count

        return DAOStats(
            totalProposals: total,
            passedProposals: passed,
            rejectedProposals: rejected,
            averageParticipation: 0,
            treasuryGrowth: 0,
            activeMembers: 0
        )
    }

    // MARK: - Helpers

    private func updateDAO(_ dao: DAO) {
        if let index = daos.firstIndex(where: { $0.id == dao.id }) {
            daos[index] = dao
            saveDAOs()
        }
    }

    private func updateProposal(_ proposal: DAOProposal, daoId: String) {
        if var daoProposals = proposals[daoId],
           let index = daoProposals.firstIndex(where: { $0.id == proposal.id }) {
            daoProposals[index] = proposal
            proposals[daoId] = daoProposals
            saveProposals()
        }
    }

    private func checkMembershipRequirements(_ dao: DAO) async throws -> Bool {
        switch dao.governance.membershipType {
        case .open:
            return true
        case .tokenGated:
            guard let tokenAddress = dao.governance.tokenAddress else { return false }
            // TODO: Check token balance
            return true
        case .nftGated:
            guard let nftAddress = dao.governance.nftAddress else { return false }
            // TODO: Check NFT ownership
            return true
        case .inviteOnly:
            return false
        }
    }

    private func generateTreasuryAddress() -> String {
        "0x" + UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(40)
    }

    private func getCurrentAddress() async throws -> String {
        // TODO: Get from WalletManager
        return "0x742d35Cc6634C0532925a3b844Bc454e4438f44e"
    }

    private func getMockMembers() -> [DAOMember] {
        [
            DAOMember(
                id: "1",
                userId: "user1",
                address: "0x742d35Cc6634C0532925a3b844Bc454e4438f44e",
                username: "Alice",
                role: .owner,
                votingPower: 10000,
                joinedAt: Date(),
                proposalsCreated: 5,
                votesCast: 12
            ),
            DAOMember(
                id: "2",
                userId: "user2",
                address: "0x8626f6940E2eb28930eFb4CeF49B2d1F2C9C1199",
                username: "Bob",
                role: .member,
                votingPower: 5000,
                joinedAt: Date(),
                proposalsCreated: 2,
                votesCast: 8
            )
        ]
    }

    // MARK: - Storage

    private func saveDAOs() {
        if let encoded = try? JSONEncoder().encode(daos) {
            UserDefaults.standard.set(encoded, forKey: storageKey)
        }
    }

    private func loadDAOs() {
        if let data = UserDefaults.standard.data(forKey: storageKey),
           let decoded = try? JSONDecoder().decode([DAO].self, from: data) {
            daos = decoded
        }
    }

    private func saveProposals() {
        if let encoded = try? JSONEncoder().encode(proposals) {
            UserDefaults.standard.set(encoded, forKey: proposalsKey)
        }
    }

    private func loadProposals() {
        if let data = UserDefaults.standard.data(forKey: proposalsKey),
           let decoded = try? JSONDecoder().decode([String: [DAOProposal]].self, from: data) {
            proposals = decoded
        }
    }
}
